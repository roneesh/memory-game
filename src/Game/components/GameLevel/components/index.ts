import LevelReady from './LevelReady';
// import LevelPlaying from './LevelPlaying';
// import LevelFinished from './LevelFinished';
// import LevelPreview from './LevelPreview';

export {
  LevelReady,
  // LevelPlaying,
  // LevelFinished,
  // LevelPreview,
};
