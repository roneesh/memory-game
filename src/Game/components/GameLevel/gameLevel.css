.game-level {
  margin-bottom: 20px;
}

.game-board {
  background-color: #4f5d75;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 3px 0;
}

.game-board__tile {
  width: 100px;
  height: 100px;
  margin: 3px;
  perspective: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #bfc0c0;
  color: #ffffff;
  font-size: 32px;
  font-family: 'Monoton', cursive;
}

@keyframes tileWiggle {
  0%   { transform:  rotateZ(0deg) }
  25%  { transform:  rotateZ(5deg) }
  50%  { transform:  rotateZ(-5deg) }
  100% { transform:  rotateZ(0deg) }
}


.game-board__tile--wiggling {
  position: relative;
  animation-name: tileWiggle;
  animation-duration: 3s;
}
