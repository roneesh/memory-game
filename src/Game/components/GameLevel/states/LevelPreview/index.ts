import LevelPreview from './LevelPreview';

export default LevelPreview;
