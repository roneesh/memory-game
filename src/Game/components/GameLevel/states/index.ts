// import LevelFinished from './LevelFinished';
// import LevelPlaying from './LevelPlaying';
import LevelPreview from './LevelPreview';
import LevelReady from './LevelReady';

export {
  // LevelFinished,
  // LevelPlaying,
  LevelPreview,
  LevelReady,
};
