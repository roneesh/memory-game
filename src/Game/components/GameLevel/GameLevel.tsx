import { shuffle } from 'lodash';
import * as React from 'react';
import './gameLevel.css';
import { LevelPreview, LevelReady, } from './states';

interface IGameLevelProps {
  updateLevel: () => void;
};

type LevelState = "ready" | "preview" | "playing" | "finished";
export interface IBoardTile {
  character: string | HTMLElement;
  visible: boolean;
};
export interface ILevelTurns {
  count: number;
  currentTurn: number[];
}
interface IGameLevelState {
  levelState: LevelState;
  board: IBoardTile[];
  turns:  ILevelTurns;
};

class GameLevel extends React.Component<IGameLevelProps, IGameLevelState> {
  constructor(props: IGameLevelProps) {
    super(props);

    this.state = {
      board: this.generateBoard(),
      levelState: 'ready',
      turns: {
        count: 0,
        currentTurn: [],
      },
    };
  }

  public generateBoard(): IBoardTile[] {
    return shuffle(
      ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
      'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'].map(character => {
        return {
          character,
          visible: false,
        };
      })
    );
  }

  public setLevelState = (levelState: LevelState) => {
    this.setState({ levelState });
  }

  public getResetBtn() {
    const { levelState } = this.state;

    const text = levelState === 'ready' ? 'Play Level' : 'Reset';
    const nextState = levelState === 'ready' ? 'preview' : 'ready';
    return <button onClick={this.setLevelState.bind(null, nextState)}>
      {text}
    </button>;
  }

  public renderControls() {

    return <div className="game-level__controls">
      {this.getResetBtn()}
    </div>
  }

  public renderState() {
    const { levelState, board } = this.state;
    switch(levelState) {
      case 'ready':
        return <LevelReady board={board} />;
      case 'preview':
        return <LevelPreview board={board} />;
      // case 'playing':
      //   return <LevelPlaying />;
      // case 'finished':
      //   return <LevelFinished />;
      default:
        return null;
    }
  }

  public render() {
    return <div className="game-level">
      {this.renderControls()}
      {this.renderState()}
    </div>;
  }
}

export default GameLevel;
