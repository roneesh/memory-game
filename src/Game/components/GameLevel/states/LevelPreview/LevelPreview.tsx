import * as React from 'react';
import { IBoardTile } from '../../GameLevel';

interface ILevelPreviewProps {
  board: IBoardTile[];
  setLevelState: (newLevelState: string) => void;
};

interface ILevelPreviewState {
  previewTimer: NodeJS.Timer;
  timeElapsed: number;
  previewedTileIndex: number;
  wiggling: boolean;
};

class LevelPreview extends React.Component<ILevelPreviewProps, ILevelPreviewState> {
  public timeIncrement = 200;

  constructor(props: ILevelPreviewProps) {
    super(props);

    this.state = {
      previewTimer: setInterval(this.previewManager, this.timeIncrement),
      previewedTileIndex: 0,
      timeElapsed: 0,
      wiggling: false,
    };
  }

  public previewManager = () => {
    const { setLevelState } = this.props;
    const { timeElapsed } = this.state;
    if (timeElapsed < 4800) {
      this.setState({
        previewedTileIndex: this.state.previewedTileIndex + 1,
        timeElapsed: timeElapsed + this.timeIncrement,
      });
      return;
    }
    if (timeElapsed < 7800) {
      this.setState({
        timeElapsed: timeElapsed + this.timeIncrement,
        wiggling: true,
      });
      return;
    }
    setLevelState('playing');
  }

  public incrementPreviewedTile = () => {
    this.setState({
      previewedTileIndex: this.state.previewedTileIndex + 1,
    })
  }

  public render() {
    const { board } = this.props;
    const { previewedTileIndex, wiggling } = this.state;
    const tileClass = `game-board__tile ${wiggling ? 'game-board__tile--wiggling' : ''}`;

    return <div className="game-board">
      {board.map((item, index) => {
        const character = index < previewedTileIndex ? item.character : null;
        return <div className={tileClass} key={index}>
          {character}
      </div>;
      })}
    </div>
  }

};

export default LevelPreview;
