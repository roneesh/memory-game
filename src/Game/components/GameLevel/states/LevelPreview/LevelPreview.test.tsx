import { shallow } from 'enzyme';
import * as React from 'react';
import renderer from 'react-test-renderer';
import { IBoardTile } from '../../GameLevel';
import LevelPreview from './LevelPreview';

describe('The level when its previewing', () => {
  let board: IBoardTile[] = [];
  let setLevelStateSpy: any;
  const boardBase =
    ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
    'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'];

  const getComponent = () => shallow((<LevelPreview board={board} setLevelState={setLevelStateSpy} />));
  const getSnapshot = () => <LevelPreview board={board} setLevelState={jest.fn()} />;

  beforeEach(() => {
    setLevelStateSpy = jest.fn();
    board = boardBase.map(character => {
      return {
        character,
      };
    });
  });

  it('it renders', () => {
    const levelPreview = renderer
      .create(getSnapshot())
      .toJSON();
    expect(levelPreview).toMatchSnapshot();
  });

  it('increments its previewedTileIndex until its time elapsed is 4800', () => {
    const component = getComponent();
    const instance = component.instance() as LevelPreview;
    jest.runTimersToTime(4800);
    expect(instance.state.previewedTileIndex).toEqual(24);
  });

  it('is wiggling from 4800ms to 7800ms', () => {
    const component = getComponent();
    const instance = component.instance() as LevelPreview;
    jest.runTimersToTime(7800);
    expect(instance.state.wiggling).toEqual(true);
    expect(component.find('.game-board__tile.game-board__tile--wiggling').length).toEqual(24);
  });

  it('dispatches the next state update after 7800ms', () => {
    getComponent();
    jest.runTimersToTime(8000);
    expect(setLevelStateSpy).toHaveBeenCalledWith('playing');
  });

  it('clears its interval on unmount', () => {
    const component = getComponent();
    const clearIntervalSpy = jest.spyOn(global, 'clearInterval');
    const interval = (component.instance() as LevelPreview).state.previewTimer;
    component.unmount();
    expect(clearIntervalSpy).toHaveBeenCalledWith(interval);
  });
});
