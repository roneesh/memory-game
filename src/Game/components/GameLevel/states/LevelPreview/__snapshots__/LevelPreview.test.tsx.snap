// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level when its previewing it renders 1`] = `
<div
  className="game-board"
>
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
  <div
    className="game-board__tile "
  />
</div>
`;
